tespy/components/customs.py

```diff
@@ -0,0 +1,818 @@
+                self.mat_deriv[k, 0, 2] = self.numeric_deriv(f, 'h', 0)
+            if not vec_z[3, 1]:
+                self.mat_deriv[k, 3, 1] = self.numeric_deriv(f, 'p', 3)
+            if not vec_z[3, 2]:
+                self.mat_deriv[k, 3, 2] = self.numeric_deriv(f, 'h', 3)
+            k += 1
+
+        ######################################################################
+        # derivatives for specified zeta at hot side 2
+        if self.zeta2.is_set:
+            f = self.zeta2_func
+            if not vec_z[1, 0]:
+                self.mat_deriv[k, 1, 0] = self.numeric_deriv(f, 'm', 1)
+            if not vec_z[1, 1]:
+                self.mat_deriv[k, 1, 1] = self.numeric_deriv(f, 'p', 1)
+            if not vec_z[1, 2]:
+                self.mat_deriv[k, 1, 2] = self.numeric_deriv(f, 'h', 1)
+            if not vec_z[4, 1]:
+                self.mat_deriv[k, 4, 1] = self.numeric_deriv(f, 'p', 4)
+            if not vec_z[4, 2]:
+                self.mat_deriv[k, 4, 2] = self.numeric_deriv(f, 'h', 4)
+            k += 1
+
+        ######################################################################
+        # derivatives for specified zeta at cold side
+        if self.zeta3.is_set:
+            f = self.zeta3_func
+            if not vec_z[2, 0]:
+                self.mat_deriv[k, 2, 0] = self.numeric_deriv(f, 'm', 2)
+            if not vec_z[2, 1]:
+                self.mat_deriv[k, 2, 1] = self.numeric_deriv(f, 'p', 2)
+            if not vec_z[2, 2]:
+                self.mat_deriv[k, 2, 2] = self.numeric_deriv(f, 'h', 2)
+            if not vec_z[5, 1]:
+                self.mat_deriv[k, 5, 1] = self.numeric_deriv(f, 'p', 5)
+            if not vec_z[5, 2]:
+                self.mat_deriv[k, 5, 2] = self.numeric_deriv(f, 'h', 5)
+            k += 1
+
+        ######################################################################
+        # derivatives for additional equations
+        self.additional_derivatives(vec_z, k)
+
+    def additional_derivatives(self, vec_z, k):
+        r"""
+        Calculates matrix of partial derivatives for
+        given additional equations.
+
+        Returns
+        -------
+        mat_deriv : ndarray
+            Matrix of partial derivatives.
+        """
+
+        ######################################################################
+        # derivatives for saturated liquid at hot side 1 outlet equation
+        if self.subcooling.val is False:
+            o1 = self.outl[0].to_flow()
+            self.mat_deriv[k, 3, 1] = -dh_mix_dpQ(o1, 0)
+            self.mat_deriv[k, 3, 2] = 1
+            k += 1
+
+        ######################################################################
+        # derivatives for saturated gas at cold side outlet 3 equation
+        if self.overheating.val is False:
+            o3 = self.outl[2].to_flow()
+            self.mat_deriv[k, 5, 1] = -dh_mix_dpQ(o3, 1)
+            self.mat_deriv[k, 5, 2] = 1
+            k += 1
```

Comment:
See comment on lines 61-65, we can move this into the derivatives part.

Reply:
Yes, move into derivatives part.